oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java

```diff
@@ -773,6 +785,122 @@ public UserManager getUserManager() throws RepositoryException {
+
+    @Nullable
+    @Override
+    public BinaryHttpUpload initiateHttpUpload(String path, long maxSize, int maxParts) throws RepositoryException {
+        if (! hasPermission(path, Session.ACTION_SET_PROPERTY)) {
+            throw new AccessDeniedException(String.format("No permission to add binary property at path %s", path));
+        }
+        return sd.safePerformNullable(new ReadOperation<BinaryHttpUpload>("initiateHttpUpload") {
+
+            @Nullable
+            @Override
+            public BinaryHttpUpload performNullable() throws RepositoryException {
+                // TODO: some form of access control check
+                //       to avoid that e.g. anonymous can add unlimited binaries to the data store
+                //       and throw AccessDeniedException if this check fails
+
+                HttpBlobProvider httpBlobProvider = getHttpBlobProvider();
+                if (httpBlobProvider == null) {
+                    return null;
+                }
+
+                BlobHttpUpload upload = httpBlobProvider.initiateHttpUpload(maxSize, maxParts);
+                if (upload != null) {
+                    return new BinaryHttpUpload() {
+                        @Override
+                        public String getUploadToken() {
+                            return upload.getUploadToken();
+                        }
+
+                        @Override
+                        public long getMinPartSize() {
+                            return upload.getMinPartSize();
+                        }
+
+                        @Override
+                        public long getMaxPartSize() {
+                            return upload.getMaxPartSize();
+                        }
+
+                        @Nonnull
+                        @Override
+                        public Collection<URL> getUploadURLs() {
+                            return upload.getUploadURLs();
+                        }
+                    };
+                }
+                return null;
+            }
+        });
+    }
+
+    @Nonnull
+    @Override
+    public Binary completeHttpUpload(String uploadToken) throws RepositoryException {
+        return sd.perform(new ReadOperation<Binary>("completeHttpUpload") {
+
+            @Nonnull
+            @Override
+            public Binary perform() throws RepositoryException {
+
+                HttpBlobProvider httpBlobProvider = getHttpBlobProvider();
+                if (httpBlobProvider == null) {
+                    throw new RepositoryException("HTTP binary upload not supported");
+                }
+                Blob blob = httpBlobProvider.completeHttpUpload(uploadToken);
+                if (blob == null) {
+                    throw new RepositoryException("HTTP binary upload could not be completed");
+                }
+
+                ValueFactoryImpl vf = (ValueFactoryImpl) sessionContext.getValueFactory();
```

Comment:
@Oak committers: Would it make sense for `SessionContext.getValueFactory()` to return `ValueFactoryImpl` right away, so we don't need the cast?

`SessionContext` always creates a `ValueFactoryImpl` in its constructor, and AFAICS `SessionContext` is an internal object that isn't required to stick to the API interfaces (but I might be missing something).

Alternatively, we could add a new interface `BlobValueFactory` that `ValueFactoryImpl` can implement, and which would provide `Binary createBinary(Blob)` and `String getBlobId(Binary)` (used below in `getHttpDownloadURL()`)..